### ISSUE TITLE
"make check-ada" failures with Xcode Command Line Tools 15.1b3

### ISSUE THREAD

**Issue by simonjwright** [iains/gcc-13-branch#10]
For GCC 13.2.0, both x86_64 and aarch64, there are many failures with CLT 15.1b*. It's OK with CLT 14.2.

This is not so for GCC 14.0.0 (I've only checked aarch64 builds).

It looks as though the failures are related to exception handling (I've certainly had a case outside the test suites where an exception that should have been caught wasn't). Evaluating this isn't helped by the fact that many of the tests raise exceptions when they detect errors.

Could it be that upstream have revised exception handling in 14.0.0 in such a way that it doesn't trigger this problem?

Results below: 
* the 13.2.0 FSF release on x86_64 with many ACATS failures, 
* the tagged release here on aarch64 with CLT 15.1b3 (with similarly many ACATS failures and a lot of GNAT failures too), 
* 14.0.0 on aarch64 with CLT 14.2 (OK), 
* 14.0.0 on aarch64 with CLT 15.1b3 (OK).

### `gnat.sum` ###
```
# of expected passes		3391
# of unexpected failures	3
# of expected failures		23
# of unsupported tests		15
```

### `acats.sum` ###
```
# of expected passes		2281
# of unexpected failures	47
*** FAILURES: c250002 c64106a c94010a c94011a c95087a cxa4027 cxf3004
cxf3a08 c46014a c940010 cb41002 cxa4008 cxa4009 c34007j c43206a
c45304a c48006b c650001 c94001a c94001b c94001c c9a007a c9a009c
cxa4019 cxa4020 c83030a c83030c c93004a c93005a cc1220a ce2402a
ce2404a ce2404b ce2405b c380003 c45632a c74004a cz1102a c45504a
c85005g c93005b c95040c cxac003 c390002 c94002a c94002f c94002g
```

### `gnat.sum` ###
```
# of expected passes		2584
# of unexpected failures	761
# of expected failures		23
# of unsupported tests		39
```

### `acats.sum` ###
```
# of expected passes		2285
# of unexpected failures	43
*** FAILURES: c250002 c43206a c650001 cxa4027 c45304a c48006b c94002a
c94002f c94002g c34007j c45504a c46014a c95040b ce2402a ce2404a
ce2404b ce2405b cxf3004 cxf3a08 c390002 c64106a c85005g c940013
c94001a c94001b c94001c c95087a c9a007a cxac003 c45632a c83030a
c83030c c94010a c94011a cb1010a cb1010d cc1220a cz1102a c380003
c74004a c940010 cb41002 cxa4008
```

### `gnat.sum` ###
```
# of expected passes		3350
# of unexpected failures	12
# of expected failures		23
# of unsupported tests		35
```

### `acats.sum` ###
```
# of expected passes		2325
# of unexpected failures	3
*** FAILURES: c250002  cb1010a cb1010d 
```

## 14.0.0 20231120 31499d aarch64 / CLT 14.2 ##
(can't build with CLT 15.1b3, see https://github.com/iains/gcc-darwin-arm64/issues/125#issuecomment-1825871563)

### `gnat.sum` ###
```
# of expected passes		3368
# of unexpected failures	2
# of expected failures		23
# of unsupported tests		35
```

### `acats.sum` ###
```
# of expected passes		2324
# of unexpected failures	4
*** FAILURES: c250002 cb1010a cb1010c cb1010d 
```

**Comment by iains** [comment]
FAOD, this is with the in-tree ACATS, not a newer version?

14.2 CLT == ld-classic?

what happens if you use the ld-classic workaround on 15.1b*?
(I did not get a chance to try 15.1b3)

My last test results on Sonoma (x86_64, XC CLT 15.1b) are:
```
LAST_UPDATED: Mon Nov 20 10:15:22 UTC 2023 (revision r14-5613-g103a3966bc7b)

                === acats tests ===

                === acats Summary ===
# of expected passes            2328
# of unexpected failures        0
Native configuration is x86_64-apple-darwin23

```

edit : that was bootstrapped with gcc-11.4

**Comment by simonjwright** [comment]
> FAOD, this is with the in-tree ACATS, not a newer version?

Yes.

> 14.2 CLT == ld-classic?

No; but CLT 14.2 doesn’t actually _have_ ld-classic.

> what happens if you use the ld-classic workaround on 15.1b*? (I did not get a chance to try 15.1b3)

That’s the fix. It hadn’t occurred to me that Apple wouldn’t have fixed `ld`!

I wondered whether the Xcode beta (3) might have fixed the problem; no. `-Wl,-ld_classic` is still required.

Considering my users, I’m wondering whether it’d be appropriate to look into a shim for `ld` which checks `xcrun --find ld-classic` & delegates to that if found, otherwise to standard `ld`?

> My last test results on Sonoma (x86_64, XC CLT 15.1b) are:
> 
> ```
> LAST_UPDATED: Mon Nov 20 10:15:22 UTC 2023 (revision r14-5613-g103a3966bc7b)
> 
>                 === acats tests ===
> 
>                 === acats Summary ===
> # of expected passes            2328
> # of unexpected failures        0
> Native configuration is x86_64-apple-darwin23
> ```
> 
> edit : that was bootstrapped with gcc-11.4

Was that with HFS? otherwise I’d have expected the C250002 failure.

----

Now that we’ve traced the problem to a known issue (albeit on gcc-darwin-arm64) do you want to close this one?

**Comment by iains** [comment]
> > what happens if you use the ld-classic workaround on 15.1b*? (I did not get a chance to try 15.1b3)
> 
> That’s the fix. It hadn’t occurred to me that Apple wouldn’t have fixed `ld`!

> > I wondered whether the Xcode beta (3) might have fixed the problem; no. `-Wl,-ld_classic` is still required.

Well, AFAIK, we thought it **_was_** fixed (although I still did not test b3)
@fxcoudert ?

> Considering my users, I’m wondering whether it’d be appropriate to look into a shim for `ld` which checks `xcrun --find ld-classic` & delegates to that if found, otherwise to standard `ld`?

Hmm .. your users are building from source, or using a toolchain you built?

> > My last test results on Sonoma (x86_64, XC CLT 15.1b) are:
> > ```
> > LAST_UPDATED: Mon Nov 20 10:15:22 UTC 2023 (revision r14-5613-g103a3966bc7b)
> > 
> >                 === acats tests ===
> > 
> >                 === acats Summary ===
> > # of expected passes            2328
> > # of unexpected failures        0
> > Native configuration is x86_64-apple-darwin23
> > ```

> Was that with HFS? otherwise I’d have expected the C250002 failure.

Yes, most of my systems have external drives for the build; it tends to hammer disks somewhat.  Some are APFS, but that one is HFS+

> Now that we’ve traced the problem to a known issue (albeit on gcc-darwin-arm64) do you want to close this one?

not sure, yet - we seem to have a potential issue with 15.1b3 ...

**Comment by simonjwright** [comment]
> > Considering my users, I’m wondering whether it’d be appropriate to look into a shim for `ld` which checks `xcrun --find ld-classic` & delegates to that if found, otherwise to standard `ld`?
> 
> Hmm .. your users are building from source, or using a toolchain you built?

Using a toolchain I built

**Comment by simonjwright** [comment]
I just rebuilt 14.0.0 with CLT 15.1b3, using a shim:
```
#!/bin/sh

classic=$(xcrun --find ld-classic 2>/dev/null) || true

if [ -n "$classic" ]; then
    exec $classic "$@"
else
    exec ld "$@"
fi
```
which resulted in:
#### gnat.sum ####
```
# of expected passes		3358
# of unexpected failures	12
# of expected failures		23
# of unsupported tests		35
```
#### acats.sum ####
```
# of expected passes		2325
# of unexpected failures	3
*** FAILURES: c250002  cb1010a cb1010d 
```

**Comment by fxcoudert** [comment]
> Well, AFAIK, we thought it was fixed (although I still did not test b3)

I'm not testing betas anymore, I don't have time. It was reported to me as fixed in 15.1 beta, but I haven't confirmed it myself.

**Comment by simonjwright** [comment]
> > Well, AFAIK, we thought it was fixed (although I still did not test b3)
> 
> I'm not testing betas anymore, I don't have time. It was reported to me as fixed in 15.1 beta, but I haven't confirmed it myself.

OK, but from where I’m sitting Apple will release 15.1 and we won’t be able to build GCC. Unless of course `libgomp.dylib` only comes into play in Ada builds?

Which version of XC/CLTs are you using?

**Comment by iains** [comment]
> > > Well, AFAIK, we thought it was fixed (although I still did not test b3)
> > 
> > 
> > I'm not testing betas anymore, I don't have time. It was reported to me as fixed in 15.1 beta, but I haven't confirmed it myself.
> 
> OK, but from where I’m sitting Apple will release 15.1 and we won’t be able to build GCC. Unless of course `libgomp.dylib` only comes into play in Ada builds?

No. libgomp is needed for a number of things - it is actually probably unused by the Ada front end as such, but comes into play with some optimisations.

> Which version of XC/CLTs are you using?

On x86_64 Sonoma XC 15.1b (first one) CLT is OK without the ld-classic work-around, and the test results are 'nominal'.
Also very short on time to try out every beta ..

**Comment by fxcoudert** [comment]
I'm worried there is a confusion over what issue we're talking about. The `ld` bug that I am aware of, and is hit in GCC bootstrap, was:

```
ld: Assertion failed: (resultIndex < sectData.atoms.size()), function findAtom, file Relocations.cpp, line 1336.
```

filed as FB13038083. This was reported to me as fixed in the Xcode 15.1 beta 1 and later.

---

If I read correctly, now you're referring to a different bug, which is hit during the build of libgomp (https://github.com/iains/gcc-darwin-arm64/issues/125), and whose symptom is:

```
ld: address=0x0 points to section(3) with no content in '/Volumes/Miscellaneous3/aarch64/14.0.0/gcc/aarch64-apple-darwin21/libgomp/.libs/target-indirect.o'
```

Is that correct? Was it filed with Apple? If not, waiting for a fix might be endless.

**Comment by simonjwright** [comment]
> ```
> ld: address=0x0 points to section(3) with no content in '/Volumes/Miscellaneous3/aarch64/14.0.0/gcc/aarch64-apple-darwin21/libgomp/.libs/target-indirect.o'
> ```
> 
> Is that correct? Was it filed with Apple? If not, waiting for a fix might be endless.

FB13416813

**Comment by simonjwright** [comment]
After configuring with `--disable-libgomp`, & with CLT 15.1b3, I had a successful build of GCC 14.0.0/31499d1 aarch64 and ran the in-tree check-ada suite; 12 failures in gnat, 3 in acats. This is slightly different from the build with CLT 14.2 reported above, but not a serious worry

**Comment by simonjwright** [comment]
Just posted FB13557418, Exception handling: linking with ld-classic is OK, ld-new fails.

Of course, I should have said "fails sometimes", will update.

**Comment by simonjwright** [comment]
The latest GCC 14’s don’t show a problem (now with XC 15.2, CLT 15.1). After a reghunt (fixhunt?) on the x86_64 snapshots I find that the fix was Iain’s patch

```
From 0fe7962afc7c01488432b98b6f442b24946a490d Mon Sep 17 00:00:00 2001
From: Iain Sandoe <iain@sandoe.co.uk>
Date: Thu, 31 Aug 2023 19:20:43 +0100
Subject: [PATCH] Darwin: Match system sections and relocs for exception
 tables.

System tools from Darwin10 onwards have moved the exceptions tables from
the __DATA segment to the __TEXT one.  They also revised the relocations
used for typeinfo.  While Darwin9 was not changed at the time, in fact the
tools there are equally happy with the revised scheme - and therefore at
present there seems no reason to special-case it.

Signed-off-by: Iain Sandoe <iain@sandoe.co.uk>

gcc/ChangeLog:

        * config/darwin-sections.def (darwin_exception_section): Move to
        the __TEXT segment.
        * config/darwin.cc (darwin_emit_except_table_label): Align before
        the exception table label.
        * config/darwin.h (ASM_PREFERRED_EH_DATA_FORMAT): Use indirect PC-
        relative 4byte relocs.
```

which I’ve successfully applied to the GCC 13.2.0 sources (that is, [my demonstrator](https://github.com/simonjwright/xcode_15_fix/tree/main/xcode_issue) using a few of the ACATS 4.1EE tests no longer reports errors).

It seems entirely reasonable that code in this area would affect detailed linker behaviour!

Would it be possible for this patch to be backported to earlier GCC branches? (I’m most interested in 13).

**Comment by iains** [comment]
> Would it be possible for this patch to be backported to earlier GCC branches? (I’m most interested in 13).

I plan to backport fixes that make our output (more) ABI compliant.  As always, time is the issue (hopefully a 14.0.1 push will come later today).

**Comment by iains** [comment]
fixed in 13.3-darwin-r0
